### ISSUE TITLE
Computing mean loss efficiently with lazy tensors

### ISSUE THREAD

**Issue by coleridge72** [pytorch/xla#3435]
## ❓ Questions and Help
Hi all

In my graph I compute the loss and take a gradient step in the standard way.
I then append the loss to a buffer which is averaged at the end of the epoch and saved on Tensorboard.

I am curious what the most efficient way to do this is without running the entire graph twice. Below when I call `.backward()` and `.cpu()` on the loss will this run the graph twice? Or is this automatically avoided?

Thanks!

pseudocode:
```
loss_buffer = []
for step in epoch:
  loss = compute_loss()
  loss.backward()
  xm.optimizer_step(opt)
  loss_buffer.append(loss.detach().cpu())

events.add_scalar('epoch_loss', torch.mean(loss_buffer))
```

(I'm using a TPU with pytorch 1.10)

**Comment by JackCaoG** [comment]
`xm.optimizer_step(opt, barrier=True)` should be enough. It will cut the graph, execute the pending graph and materialized all tensor to a concrete value. Next time you access any live tensor(loss in this case) it will just do a device side fetch instead of execute the graph. 

You can take a look at https://github.com/pytorch/xla/blob/c74deb63385d7d5c5db3312629a7adc83d075745/torch_xla/core/xla_model.py#L833

**Comment by coleridge72** [comment]
Ah ok thanks, would this slow down training compared to not reporting the loss/using the barrier?

I was thinking of recording loss at a less frequent interval but maybe this would not make any difference.

**Comment by JackCaoG** [comment]
generally speaking larger graph will result in better performance since it will give xla compiler more room to optimize. This is model by model through, I would encourage you to try both and see if it makes any difference for you in the e2e time.
